File: Design-Patterns-Using-JavaScript/Visitor/VisitorPattern.js
// Define Element interface
class Element {
    accept(visitor) {
        throw new Error("accept method must be implemented");
    }
}

// Define ConcreteElementA implementing Element
class ConcreteElementA extends Element {
    accept(visitor) {
        visitor.visitElementA(this);
    }
}

// Define ConcreteElementB implementing Element
class ConcreteElementB extends Element {
    accept(visitor) {
        visitor.visitElementB(this);
    }
}

// Define Visitor interface
class Visitor {
    visitElementA(elementA) {
        throw new Error("visitElementA method must be implemented");
    }

    visitElementB(elementB) {
        throw new Error("visitElementB method must be implemented");
    }
}

// Define ConcreteVisitor1 implementing Visitor
class ConcreteVisitor1 extends Visitor {
    visitElementA(elementA) {
        console.log("ConcreteVisitor1 visitElementA() method called.");
    }

    visitElementB(elementB) {
        console.log("ConcreteVisitor1 visitElementB() method called.");
    }
}

// Define ConcreteVisitor2 implementing Visitor
class ConcreteVisitor2 extends Visitor {
    visitElementA(elementA) {
        console.log("ConcreteVisitor2 visitElementA() method called.");
    }

    visitElementB(elementB) {
        console.log("ConcreteVisitor2 visitElementB() method called.");
    }
}

// Client code
const visitor1 = new ConcreteVisitor1();
const elementA = new ConcreteElementA();
elementA.accept(visitor1);

const elementB = new ConcreteElementB();
elementB.accept(visitor1);

/*
ConcreteVisitor1 visitElementA() method called.
ConcreteVisitor1 visitElementB() method called.
*/

File: Design-Patterns-Using-Kotlin/Visitor/VisitorPattern.kt
interface Element {
    fun accept(visitor: Visitor)
}

class ConcreteElementA : Element {
    override fun accept(visitor: Visitor) {
        visitor.visitElementA(this)
    }
}

class ConcreteElementB : Element {
    override fun accept(visitor: Visitor) {
        visitor.visitElementB(this)
    }
}

interface Visitor {
    fun visitElementA(elementA: ConcreteElementA)
    fun visitElementB(elementB: ConcreteElementB)
}

class ConcreteVisitor1 : Visitor {
    override fun visitElementA(elementA: ConcreteElementA) {
        println("ConcreteVisitor1 visitElementA() method called.")
    }

    override fun visitElementB(elementB: ConcreteElementB) {
        println("ConcreteVisitor1 visitElementB() method called.")
    }
}

class ConcreteVisitor2 : Visitor {
    override fun visitElementA(elementA: ConcreteElementA) {
        println("ConcreteVisitor2 visitElementA() method called.")
    }

    override fun visitElementB(elementB: ConcreteElementB) {
        println("ConcreteVisitor2 visitElementB() method called.")
    }
}

// Client code
fun main() {
    val visitor1: Visitor = ConcreteVisitor1()
    val elementA: Element = ConcreteElementA()
    elementA.accept(visitor1)

    val elementB: Element = ConcreteElementB()
    elementB.accept(visitor1)
}

/*
ConcreteVisitor1 visitElementA() method called.
ConcreteVisitor1 visitElementB() method called.
*/